src/basic/time-util.c

```diff
@@ -874,6 +879,71 @@ int parse_timestamp(const char *t, usec_t *usec) {
         return 0;
 }
 
+typedef struct ParseTimestampResult {
+        usec_t usec;
+        int return_value;
+} ParseTimestampResult;
+
+int parse_timestamp(const char *t, usec_t *usec) {
+        char *last_space, *timezone = NULL;
+        ParseTimestampResult *shared, tmp;
+        int r;
+        pid_t pid;
+
+        last_space = strrchr(t, ' ');
+
+        if (last_space != NULL) {
```

Comment:
Hmm, you have the exact same code in `calendar_spec_next_usec` above. This needs to be extracted into a helper function, not repeated.

Reply:
@keszybz I was not sure whether it makes sense to have it implemented first, then refactored as a part of the next PR.

Reply:
@keszybz they are vaguely similar. It duplicates statements, not the knowledge. (so it's not a DRY violation)

Reply:
They are sufficiently similar to unify the common part: lines 1362–1400.

Reply:
Not sure how nice it would look like though: they are different in what main "implementation" they call. They are still 2 different implementations that happen to have some parts similar (since they do similar work). This similarity is accidental, and there is no relationship between them to expect them to evolve identically.

I will try to extract it to a shared common function, but its use would always be limited by these 2 cases only.

I personally would not do that, but let's try and see.

Reply:
@keszybz I spent some time trying and I don't see how it's possible to generalise it in a way that does not make it 10 times more complicated spaghetti of functions and macroses.

Reply:
OK, so please push the best version you have without this change, and we'll go from there.

Reply:
@keszybz what currently is commited contains all the changes requested from you and Lennart.